Fix: Binding doesn't respect special IValueConverter return values

bakerhillpins opened the pull request:
### Description of Change

Update Binding classes in MAUI to support special return values from the `IValueConverter` assigned to the `Binding.Converter` property.

Make changes to stop Bindings from being resolved multiple times during the course of a single update or change to the children of the View.

I've modified the MAUI source to support the both the Binding.DoNothing and the BindableProperty.UnsetValue functionality described in the [IValueConverter](https://learn.microsoft.com/en-us/dotnet/api/system.windows.data.ivalueconverter.convert?view=windowsdesktop-7.0#remarks) and [IMultiValueConverter](https://learn.microsoft.com/en-us/dotnet/api/system.windows.data.imultivalueconverter.convert?view=windowsdesktop-7.0&viewFallbackFrom=net-6.0#returns) docs. These changes support both the [Binding Fallbacks](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/data-binding/binding-fallbacks?view=net-maui-7.0) and the [MultiBinding](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/data-binding/multibinding?view=net-maui-6.0) discussions in the docs and this would then operate as WPF does.

The one omission here is it's unclear how the [StringFormat](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/data-binding/string-formatting?view=net-maui-7.0) behavior should be handled as it's not specifically discussed anywhere in MAUI or WPF documentation that I can find. Discussions on StackOverflow indicate that WPF bindings ignore StringFormat during application of TargetNullValue or FallbackValue and so I'm assuming that MAUI should keep this and I've made changes to support that.

It should also be noted that [IValueConverter.ConvertBack](https://learn.microsoft.com/en-us/dotnet/api/system.windows.data.ivalueconverter.convertback?view=windowsdesktop-7.0) discusses using Fallback and Targetnull Values however the [documentation](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/data-binding/binding-fallbacks?view=net-maui-6.0#define-a-null-replacement-value) for these clearly states it's applied to Source values only:

`The TargetNullValue property allows a replacement value to be defined that will be used when the binding source is resolved, but the value is null.`

I originally targeted the NET6.0 branch because the documentation (until yesterday morning) specifically said to do this if you expected this to be applied to NET6. I have that branch available if needed.

### Issues Fixed

fixes #11740 
fixes #10792
fixes #10806

bakerhillpins committed:
Add UnitTests supporting Binding return values.

Unit tests that support Binding IValueConverter converters that return Binding.DoNothing and BindableProperty.UnsetValue.

bakerhillpins committed:
Add unit tests for IValueConverter.ConvertBack return values.

bakerhillpins committed:
Update Bindings to support "special" return values. 

This is a first pass at updating the Binding classes to support the IValueConverter and IMultiValueConverter as spec'd on the desktop. Issues remain with over processing these return values, for example they can get picked up by StringFormat when they should not.

bakerhillpins committed:
Additional support and fixes for IValueConverter return values.

Fix: Allow FallbackValue to be replaced with default value where necessary.
Process return values from GetSourceValue.
Don't update MultiBindings unless there's been a change in one of the included bindings.

bakerhillpins committed:
Support IValueConverter.ConvertBack return values.

Use same implementation as MultiBinding. IValueConverter.ConvertBack talks about using FallbackValue and NulltargetValue with the special return values but the documentation clearly states that those 2 properties on a Binding are only applied to Source value operations.

rmarinho commented:
/azp run

bakerhillpins committed:
Create ApplyCore method

DRY - use common method for Apply to remove duplicated code.

bakerhillpins committed:
Optimize TypedBinding resolution from source

Rework flow so that FallbackValue is only calculated for a from source Binding operation when it's needed (rather than as a default value initialization).

bakerhillpins commented:
Hope I didn't miss the pipeline run. I made some changes for DRY and optimization of `TypedBinding`.  I realize that performance measurements are VERY subjective but I ran a few cycles on both the original code (main branch) and on my most recent updates to be sure that I wasn't mucking up the works. This is what I got for your reference. I'd be interested to know what you folks see if anything.

fix-11740 branch:
```
      SpeedTestApply Failed: Applying 100000 Typedbindings			: 29ms.
Xunit.Sdk.XunitException
Applying 100000 Typedbindings			: 29ms.
Applying 100000 Typedbindings (without INPC)	: 16ms.
Applying 100000 Bindings			: 82ms.
Setting  100000 values				: 7ms.
   at Microsoft.Maui.Controls.Core.UnitTests.TypedBindingUnitTests.SpeedTestApply() in D:\Repos\Forks\maui\src\Controls\tests\Core.UnitTests\TypedBindingUnitTests.cs:line 1999
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)


      SpeedTestSetBC Failed: Setting BC for 100000 Typedbindings			: 116ms.
Xunit.Sdk.XunitException
Setting BC for 100000 Typedbindings			: 116ms.
Setting BC for 100000 Typedbindings (without INPC)	: 92ms.
Setting BC for 100000 Bindings				: 154ms.
Setting  100000 values					: 7ms.
   at Microsoft.Maui.Controls.Core.UnitTests.TypedBindingUnitTests.SpeedTestSetBC() in D:\Repos\Forks\maui\src\Controls\tests\Core.UnitTests\TypedBindingUnitTests.cs:line 2065
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```
main branch:
```
      SpeedTestApply Failed: Applying 100000 Typedbindings			: 34ms.
Xunit.Sdk.XunitException
Applying 100000 Typedbindings			: 34ms.
Applying 100000 Typedbindings (without INPC)	: 18ms.
Applying 100000 Bindings			: 93ms.
Setting  100000 values				: 9ms.
   at Microsoft.Maui.Controls.Core.UnitTests.TypedBindingUnitTests.SpeedTestApply() in D:\Repos\Forks\maui\src\Controls\tests\Core.UnitTests\TypedBindingUnitTests.cs:line 1655
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)


      SpeedTestSetBC Failed: Setting BC for 100000 Typedbindings			: 129ms.
Xunit.Sdk.XunitException
Setting BC for 100000 Typedbindings			: 129ms.
Setting BC for 100000 Typedbindings (without INPC)	: 113ms.
Setting BC for 100000 Bindings				: 171ms.
Setting  100000 values					: 9ms.
   at Microsoft.Maui.Controls.Core.UnitTests.TypedBindingUnitTests.SpeedTestSetBC() in D:\Repos\Forks\maui\src\Controls\tests\Core.UnitTests\TypedBindingUnitTests.cs:line 1721
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)

```

mattleibow commented:
Adding the perf label as well as this does have some exciting changes. @jonathanpeppers not sure if you are interested in checking this as well?

mattleibow commented:
/azp run

bakerhillpins commented:
FYI, I understand the problem associated with #10792 and have a fix.  I believe I understand #10806 as well but still investigating.  May be the same fix for both.

StephaneDelcroix commented:
does this cover compiled bindings too ?

bakerhillpins commented:
> does this cover compiled bindings too ?

yes

jonathanpeppers commented on src/Controls/src/Core/MultiBinding.cs:
```diff
@@ -18,6 +18,7 @@ public sealed class MultiBinding : BindingBase
 		BindableObject _proxyObject;
 		BindableProperty[] _bpProxies;
 		bool _applying;
+		private bool _updates;
```
Can we give this variable a more descriptive name? We check if it is `false` and return early in some cases, but it's not clear why.

bakerhillpins replied:
Agreed. I'll get that fixed shortly. Just an oversight.

bakerhillpins replied:
Changed. Let me know if that's not going to work for you.

jonathanpeppers commented on src/Controls/src/Core/MultiBinding.cs:
```diff
@@ -153,26 +119,19 @@ internal override void Apply(object context, BindableObject targetObject, Bindab
 			}
+			_applying = false;
+
+			if (this.GetRealizedMode(_targetProperty) == BindingMode.OneWayToSource || !_updates )
+				return;
```
If `_updates` had a clearer name, maybe others would understand why we are returning early here.

bakerhillpins replied:
See above.

jonathanpeppers reviewed (commented):
For all performance changes, we should add a benchmark here:

https://github.com/dotnet/maui/tree/main/src/Core/tests/Benchmarks

A new benchmark comparing the different binding modes, would be great!

This way we have numbers for *how much* this improves things. And we will know it for sure won't be worse on accident!

bakerhillpins commented:
@jonathanpeppers 

There's room for more DRY changes. Specifically putting a subclass in between `BindingBase` and several of it's inheritors. This class would carry the common code for the `Converter` and `ConverterParameter` properties and the associated `GetSourceValue`/`GetTargetValue` methods which are identical. I wasn't sure if this was appropriate because someone [commented out the call](https://github.com/dotnet/maui/blob/1c4c0dd17b0cd51cedc4883b3e5a683e453685e9/src/Controls/src/Core/TypedBinding.cs#L173) to the base class code in `TypedBinding` so I figured it was a change left out for performance reasons.

Should the subclass be added?  This change would apparently also remove duplicate documentation code.

bakerhillpins committed:
Update variable name for clarity.

bakerhillpins committed:
Use existing converters for tests

Remove the converters that were added as they duplicated functionality that already exists in existing implementation.  The TestConverterReturn class isn't actually needed.

bakerhillpins committed:
Fix-10792 Multibindings applied to BindingContext property do not work.

Coalesce a context value from multiple sources as is done in Binding.Apply.  Include appropriate tests.

bakerhillpins commented:
Add fix and tests for #10792

bakerhillpins committed:
Fix-10806 Apply bindings twice when child added

Apply bindings is called as a result of setting the child.Parent on the previous line so remove the duplicate call.

bakerhillpins committed:
Fix-10806 Double apply when BindingContext has binding.

Don't call ApplyBindings and OnBindingContextChanged directly when a binding is active on the BindingContext property. Any update to the BindingContext Property will result in those calls if the property actually changes and local bindings need updated.

bakerhillpins committed:
Fix-10806 Apply Bindings in order to reduce multiple resolution.

Sort the BindingContext BindablePropertyContext to the front of the array if it's included. This allows it to be easily skipped if necessary and forces the BindingContext to be updated before local bindings. Thus they are only resolved a single time.

bakerhillpins committed:
Fix-10806 Unapply OneTime bindings on local set

The only difference between OneTime and OneWay bindings are the # of times they execute, so OneTime should also be removed with OneWay when the BindingContext Property is manually set.

bakerhillpins commented:
Added fix and tests for #10806

bakerhillpins committed:
Fix-10806 Cleanup

Filter out null bindings before sort in ApplyBindings (like previous)
Remove dead code, use null coalescing and pattern matching.

bakerhillpins commented:
@StephaneDelcroix @jonathanpeppers @mattleibow 
I've completed the work I wanted perform here. Please let me know if you have any questions. I'll make myself available for voice discussions if helpful/necessary.  I think these additions and fixes will help the platform.

jfversluis commented:
/azp run

bakerhillpins commented:
@jfversluis @StephaneDelcroix  I can't express how disappointed I am that this is still waiting for a single reviewer. I'd happily update to resolve the conflicts that are now blocking.

jonathanpeppers commented:
@bakerhillpins you didn't address my comment, if you could add a benchmark if this supposedly improves performance: https://github.com/dotnet/maui/pull/12060#pullrequestreview-1221087735

By *how much* does it improve things?

There is also a merge conflict, if you'd like to address that. Thanks.

bakerhillpins commented:
@jonathanpeppers  This PR is not and never was about changes for the sake of performance. The performance code mentioned is commented out in the MS source. I simply [ran it to see that I hadn't regressed](https://github.com/dotnet/maui/pull/12060#issuecomment-1351569925) and [mentioned that in my response to you](https://github.com/dotnet/maui/pull/12060#issuecomment-1355197415).  The issues resolved are discussed in the PR's initial post. 

I did however change the order of binding processing to fix an issue where bindings on an object were being evaluated 2x because property bindings were being evaluated before the BindingContext binding was on an object. Thus the bindings on that object would be evaluated a 2nd time when the BindingContext was applied. This included children bindings. #10806

> There is also a merge conflict, if you'd like to address that. Thanks.

This has been sitting here for months without conflict waiting for review/comment at the beginning of which time [I asked for feedback and received nothing](https://github.com/dotnet/maui/pull/12060#issuecomment-1370914041). As I mentioned, I'm willing to take the time to fix the conflict but right now I don't feel like anyone is taking my time seriously.

bakerhillpins commented:
> @bakerhillpins but we don't have a BenchmarkDotNet benchmark of multi-binding. I would want to see the results of that before merging this.

Why? I made NO PERFORMANCE changes.

> Your description here seems like you did something manual instead of using BenchmarkDotNet:

You don't appear to be reading what I'm writing and I don't know how to tell you any more clearly.  
### The "perfomance code" is in your tests and commented out. I didn't write it, I simply used it as a sanity check.

bakerhillpins committed:
Merge remote-tracking branch 'origin/main' into fix-11740

Conflicting files
src/Controls/src/Core/BindableObject.cs

berhir commented:
What is the status of this PR?
I ran today into the issue with `Binding.DoNothing` and found this PR that was created a long time ago.

I can feel the pain and frustration of the creator who obviously put a lot of time and thinking into this PR.
The PR fixes 3 issues and now it was marked as stale... why is no one reviewing it? @jfversluis @mattleibow @samhouts

bakerhillpins commented:
> What is the status of this PR? I ran today into the issue with `Binding.DoNothing` and found this PR that was created a long time ago.
> 
> I can feel the pain and frustration of the creator who obviously put a lot of time and thinking into this PR. The PR fixes 3 issues and now it was marked as stale... why is no one reviewing it? @jfversluis @mattleibow @samhouts

@berhir 
First off, thank you for recognizing my frustration. As for the PR, It's basically dead. I put many hours (including personal time) into this and they have zero interest. And yes I'm bent over it. The goalposts kept moving during the process and I'm done chasing, It appears they've moved yet again based upon the "Some Checks" section. 

MS puts on a good show with open source and asking for PRs/community development but they don't appear to back it up. I doubt it's a personal thing, based upon the release cycle and quality I'm seeing they are pushing too hard/fast. It's the early days of .NET all over again, unfortunately. Though YMMV.

bakerhillpins commented:
> I'm just a passerby interested in #10806 being fixed.
> 
> @bakerhillpins: Your PR is exciting! I read the posts here and it looks to me that @jonathanpeppers just wants to know that this PR does not introduce a performance regressions as `src/Controls/src/Core/BindableObject.cs` is an important piece of MAUI.

Thank you for noticing.  

From my perspective it's all bug fixes and always was.. MS made an assumption it was performance related because I ran some of MS's own "performance" code and randomly labeled it as [perf related](https://github.com/dotnet/maui/pull/12060#issuecomment-1354982993).  It would seem that no one read the 3 referenced bug reports that were fixed to determine what the PR was about.

> Maybe a benchmark similar to this one [#17756 (comment)](https://github.com/dotnet/maui/pull/17756#issue-1920318390) would be sufficient to persuade people that no such regression occurred. Maybe @albyrock87 can share some benchmarking code to make it easier to measure it.

Not going to be me.

> One additional thought is that if this PR can be split into multiple smaller PRs, it might be easier to review it. 1200 LOC is big.

Most of the changes are TESTS that I added to pick up on all the missing functionality and bugs I found in the binding code.

> Not trying to offend you even more, just sharing my honest opinion.

Thank you for that. I as well.

Alex-Dobrynin commented:
@jfversluis @jonathanpeppers @samhouts @mattleibow 
When will this be reviewed and merged? a lot of people experiencing issue described above. You just ignore creator of this PR. In the future nobody wants to spend any second of his life helping you fix your bugs with such approach and relationship to the community

jbtdevgit commented:
It's been a year, how is this not fixed yet? I guess this team won't move until a 'customer' is going to smash them with issues that they have to fix it asap.

albyrock87 commented:
Hey @bakerhillpins ,
my PR #24553 has been merged and it includes a slightly different fix for #10806 compared to what you have here.

As you can see I tried at all costs to avoid `OrderBy` to sort properties for two reasons:
1. Performance
2. It changes the order used by the developer to set bindings on the target causing potential unwanted side effects on interdependent properties

I wanted to help by preparing a PR targeting your repo to merge `main` into your branch and then solve the conflicts I caused.
Unfortunately I failed, because in the mean time `SetterSpecificity` came into the game within binding, so your code appears to not be compatible with current state of `main`.

Maybe I'm missing something, but I wanted to tell you that I tried :) and what was the outcome.

Thanks for all the work you've done here.

bakerhillpins commented:
> Hi @bakerhillpins. We have added the "s/pr-needs-author-input" label to this issue, which indicates that we have an open question/action for you before we can take further action. This PRwill be closed automatically in 14 days if we do not hear back from you by then - please feel free to re-open it if you come back to this PR after that time.

LOL, interesting, I don't see any open question/action?

Classic... looks like MS automation needs some work too.
![image](https://github.com/user-attachments/assets/eb54ece4-a13a-40b5-b064-c0c9b9ff2f43)

bakerhillpins commented:
> Hey @bakerhillpins , my PR #24553 has been merged and it includes a slightly different fix for #10806 compared to what you have here.
> 
> As you can see I tried at all costs to avoid `OrderBy` to sort properties for two reasons:
> 
>     1. Performance
> 
>     2. It changes the order used by the developer to set bindings on the target causing potential unwanted side effects on interdependent properties
> 
> 
> I wanted to help by [preparing a PR targeting your repo to merge `main` into your branch](https://github.com/bakerhillpins/maui/pull/1) and then solve the conflicts I caused. Unfortunately I failed, because in the mean time `SetterSpecificity` came into the game within binding, so your code appears to not be compatible with current state of `main`.
> 
> Maybe I'm missing something, but I wanted to tell you that I tried :) and what was the outcome.
> 
> Thanks for all the work you've done here.

@albyrock87 
Appreciate you taking the time you did here! Also glad you found the sorting problem. Didn't catch that in my tests. I stepped away from MAUI after this debacle so I don't know how that change effects the merging of the other fixes this PR has.

Sad that your PR was able to get a review form @StephaneDelcroix in days and I asked for MONTHS for the review and got nothing.
